lib/gecko.js

```diff
@@ -0,0 +1,207 @@
+  path.resolve(os.tmpdir(), 'gecko_server_guard.lock'),
+  {timeout: 5, tryRecovery: true}
+);
+
+
+class GeckoProxy extends JWProxy {
+  async proxyCommand (url, method, body = null) {
+    if (this.didProcessExit) {
+      throw new errors.InvalidContextError(
+        `'${method} ${url}' cannot be proxied to Gecko Driver server because ` +
+        'its process is not running (probably crashed). Check the Appium log for more details');
+    }
+    return await super.proxyCommand(url, method, body);
+  }
+}
+
+class GeckoDriverProcess {
+  constructor (opts = {}) {
+    for (const optName of ['noReset', 'verbosity', 'androidStorage']) {
+      this[optName] = opts[optName];
+    }
+    this.port = opts.systemPort;
+    this.proc = null;
+  }
+
+  get isRunning () {
+    return !!(this.proc?.isRunning);
+  }
+
+  async init () {
+    if (this.isRunning) {
+      return;
+    }
+
+    if (!this.port) {
+      await GECKO_SERVER_GUARD(async () => {
+        const [startPort, endPort] = GECKO_PORT_RANGE;
+        try {
+          this.port = await findAPortNotInUse(startPort, endPort);
+        } catch (e) {
+          throw new Error(
+            `Cannot find any free port in range ${startPort}..${endPort}. ` +
+            `Double check the processes that are locking ports within this range and terminate ` +
+            `these which are not needed anymore or set any free port number to the 'systemPort' capability`);
+        }
+      });
+    }
+
+    let driverBin;
+    try {
+      driverBin = await fs.which(GD_BINARY);
+    } catch (e) {
+      throw new Error(`${GD_BINARY} binary cannot be found in PATH. ` +
+        `Please make sure it is present on your system`);
+    }
+    const args = [
+      '-p', this.port,
+    ];
+    if (_.toLower(this.verbosity) === VERBOSITY.DEBUG) {
+      args.push('-v');
+    } else if (_.toLower(this.verbosity) === VERBOSITY.TRACE) {
+      args.push('-vv');
+    }
+    if (this.noReset) {
+      args.push('--connect-existing');
+    }
+    if (this.androidStorage) {
+      args.push('–-android-storage', this.androidStorage);
+    }
+    this.proc = new SubProcess(driverBin, args);
+    this.proc.on('output', (stdout, stderr) => {
+      const line = stdout || stderr;
+      log.debug(`[${GD_BINARY}] ${line}`);
+    });
+    this.proc.on('exit', (code, signal) => {
+      log.info(`${GD_BINARY} has exited with code ${code}, signal ${signal}`);
+    });
+    log.info(`Starting '${driverBin}' with args ${JSON.stringify(args)}`);
+    await this.proc.start(0);
+  }
+
+  async stop () {
+    if (this.isRunning) {
+      await this.proc.stop('SIGTERM');
+    }
+  }
+
+  async kill () {
+    if (this.isRunning) {
+      try {
+        await this.proc.stop('SIGKILL');
+      } catch (ign) {}
+    }
+  }
+}
+
+const RUNNING_PROCESS_IDS = [];
+process.once('exit', () => {
+  for (const pid of RUNNING_PROCESS_IDS) {
+    const command = system.isWindows() ? `taskkill.exe /PID ${pid} /F` : `kill ${pid}`;
```

Comment:
`/F` should be a fallback method on Windows. The `/F` sometimes shows an error dialog which requires GUI interactions on Windows. (Not always on my experience)
So, try without `/F` -> but the process still remains after a few seconds -> then calls `/F` is pretty better on Windows.

Reply:
I'll remove it for now completely. will add later if we have any issues there